### PR TITLE
feat: Block reference support in HTML

### PR DESCRIPTION
**Pull request by SeriousBug**
This pull request adds support for block references in HTML. It includes support for:

* Single block anchor references `![[note#^anchor]]`
* Range block anchor references `![[note#^anchor1:#^anchor2]]`
* Same file block references `![[#^anchor]]` and `![[#header]]`

Range block anchor references are allowed to be mixed with headers too, so it's possible to do `![[#header:#^anchor]]` or `![[#^anchor:#header]]` to reference only a part of a section.

It's also possible to reference single items in lists, or parts of lists. Consider the following list

```
* Aut id architecto quia.
* Sapiente sed accusamus eum.
  * Ullam optio est quia. ^one
* Reprehenderit doloribus.
  * Sint minus fuga omnis non. ^two
  * Iure neque alias dolorem. ^three
```

The reference `![[#^one]]` would only reference the list element it's next to. `![[#^two:#^three]]` would reference both elements, and anything in between.

An interesting case is something like `![[#^one:#^two]]` where the referenced items are nested in different levels of the list. In this case, the reference refers to anything after the start anchor and before the end anchor, and also the parent of the start anchor up to a common ancestor with the end. In other words, the output looks like:

```
* Sapiente sed accusamus eum.
  * Ullam optio est quia. ^one
* Reprehenderit doloribus.
  * Sint minus fuga omnis non. ^two
```

This allows you to refer to regions of lists across different depths while preserving the required context for the items you selected within the region, and avoids ending up with a broken list.

### CODE REVIEW

File: packages/engine-test-utils/src/__tests__/engine-server/markdown/noteRefv2.spec.ts [hunk 1]
```diff
@@ -684,11 +742,836 @@ describe("noteRefV2", () => {
...
+            "Minima expedita vero.",
+            "Iure neque alias dolorem."
+          );
+        },
+      },
+    });
+
+    runAllTests({
+      name: "compile",
+      testCases: [
```
Comment (kevinslin) [inline packages/engine-test-utils/src/__tests__/engine-server/markdown/noteRefv2.spec.ts:1556]: can you also add a test case for what happens when we have a nested block reference link?

Comment (kevinslin) [reply packages/engine-test-utils/src/__tests__/engine-server/markdown/noteRefv2.spec.ts:1556]: also a case when a block reference link is referencing to an invalid anchor

File: packages/engine-test-utils/src/__tests__/engine-server/markdown/remarkUtils.spec.ts [hunk 2]
```diff
@@ -5,27 +5,61 @@ import {
...
+        }
+      );
+    });
+    test("one block anchor", async () => {
+      await runEngineTestV5(
+        async ({ engine }) => {
+          const body = engine.notes["foo"].body;
+          debugger;
+          const out = RemarkUtils.findAnchors(body);
+          expect(out).toMatchSnapshot("bond");
```
Comment (kevinslin) [inline packages/engine-test-utils/src/__tests__/engine-server/markdown/remarkUtils.spec.ts:49]: "bond" is a debug value to help quickly find a given snapshot. you can take a snapshot without parameters. if you see other snapshots using this value, please remove them

**Review by kevinslin — CHANGES_REQUESTED** [review]

File: packages/engine-test-utils/src/__tests__/engine-server/markdown/noteRefv2.spec.ts (hunk 1 above)
Comment (SeriousBug) [reply packages/engine-test-utils/src/__tests__/engine-server/markdown/noteRefv2.spec.ts:1556]: > can you also add a test case for what happens when we have a nested block reference link?

Do you mean have a note with a reference to another note, that also has a reference? E.g. parsed note has `![[note1#^start:#^end]]`, then note1 has `![[note2#^start:#^end]]`?

**Comment by SeriousBug** [comment]
@kevinslin Added the tests you suggested, and caught a bug in the process.

File: packages/engine-server/src/markdown/remark/noteRefsV2.ts [hunk 3]
```diff
@@ -375,11 +377,150 @@ export function convertNoteRefASTV2(
...
+  };
+  let end: FindAnchorResult = null;
+
+  if (anchorStart) {
+    start = findAnchor({
+      nodes: bodyAST.children,
+      match: anchorStart,
+    });
+    if (_.isNull(start)) {
+      debugger;
```
Comment (kevinslin) [inline packages/engine-server/src/markdown/remark/noteRefsV2.ts:465]: remove debugger

**Review by kevinslin — CHANGES_REQUESTED** [review]
nice! yeah, i remember running into this when doing the original note ref impl. just remove the debug statement and we should be good to go

**Comment by SeriousBug** [comment]
@kevinslin Done, thanks!

**Review by kevinslin — APPROVED** [review]

**Pull request merged**